cupy/cuda/memory.pyx

```diff
@@ -188,11 +199,21 @@ cdef class UnownedMemory(BaseMemory):
             runtime._ensure_context()
             ptr_attrs = runtime.pointerGetAttributes(ptr)
             device_id = ptr_attrs.device
+            if device_id == runtime.cudaCpuDeviceId:
+                # this happens with SystemMemory...
+                device_id = device.get_device_id()
         self.size = size
         self.device_id = device_id
         self.ptr = ptr
         self._owner = owner
 
+    def __dealloc__(self):
+        # Note: Cannot raise in the destructor! (cython/cython#1613)
+        if _is_sam_enabled:
+            # we don't own the memory, we must sync before free to avoid any
+            # race condition
+            runtime.streamSynchronize(stream_module.get_current_stream_ptr())
+
```

Comment:
Q: @rongou this is new to me. How would this destructor be triggered?

Reply:
Unowned memory is not pooled right? so it should trigger once the gc kicks in but not 100% sure

Reply:
Right, but for the purpose of this PR we don't use `UnownedMemory` so I am confused..

Reply:
I think this may be due to an external library interaction, probably RMM interop?

Reply:
Yeah I feel this change was to mirror that from RMM: https://github.com/rapidsai/rmm/pull/1655 but this is problematic because if [RMM uses `UnownedMemory`](https://github.com/rapidsai/rmm/blob/ab6e2961d7b8f833f688775e941c4e2ed2bd4d8a/python/rmm/rmm/allocators/cupy.py#L39) then the fast path would not be triggered (https://github.com/cupy/cupy/pull/8442#discussion_r1774375148).

Reply:
Also this destructor would cause unconditional synchronization when CUPY_ENABLE_HMM=1 for all unowned memory, regardless of its source (i.e. whether it's system memory, or whether it comes from RMM/RAPIDS). I think it causes more harm than the perf it gains.

Reply:
fixed in [73d0fe8](https://github.com/cupy/cupy/pull/8442/commits/73d0fe85d5ad53ff5c7123baf6648f205c7113e9)

Reply:
Yes RMM uses UnownedMemory. We probably should change that.

I think I originally added this because some cupy allocated buffers were causing problems. Later ran into issue with cuml c++ allocated buffers causing similar problems. Now that we synchronize the stream in rmm, we don't need to synchronize here any more.